walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -35,106 +41,223 @@ public int compare(Event a, Event b) {
+     */
+    private HashMap<String, ArrayList<TimeRange>> getOptionalAttendeesFreeTimes(
+        HashSet<String> optionalAttendees) {
+      HashMap<String, ArrayList<TimeRange>> times = new HashMap<String, ArrayList<TimeRange>>();
+      for (String attendee : optionalAttendees) {
+        HashSet<String> attendeeSet = new HashSet<String>();
+        attendeeSet.add(attendee);
 
-  /**
-   * Adds range to ranges if it is long enough to fit in a meeting.
-   *
-   * @param range the range being considered
-   * @param ranges the list of ranges >= meetingDuration
-   * @param meetingDuration the duration of meeting to be scheduled
-   */
-  private static void addIfLongEnough(
-      TimeRange range, List<TimeRange> ranges, long meetingDuration) {
-    if (range.duration() >= meetingDuration) {
-      ranges.add(range);
+        // Find all possible meeting times for just this one attendee. Must do this to deal with
+        // double bookings.
+        times.put(attendee, getMandatoryAttendeesMeetingTimes(attendeeSet));
+      }
+      return times;
     }
-  }
 
-  /**
-   * Returns only those events that are attended by at least one attendee that is attending the
-   * meeting we are trying to schedule. More intuitively, an event is "relevant" if it is attended
-   * by at least one "relevant" attendee.
-   *
-   * @param requestAttendees the set of attendees attending the meeting ("relevant" people)
-   */
-  private static ArrayList<Event> getRelevantEvents(
-      HashSet<String> relevantAttendees, ArrayList<Event> events) {
-    ArrayList<Event> relevantEvents = new ArrayList<Event>();
-    for (Event event : events) {
-      boolean isRelevant = false;
-      for (String person : event.getAttendees()) {
-        if (relevantAttendees.contains(person)) {
-          isRelevant = true;
-          break;
+    /**
+     * Returns a change log of the number of available optional attendees over time. First part of a
+     * sweep-line algorithm.
+     *
+     * @param optionalAttendeesFreeTimes mapping of all attendees to their free times
+     */
+    private TreeMap<Integer, Integer> getChangesInOptionalAttendeesAttendance(
+        HashMap<String, ArrayList<TimeRange>> optionalAttendeesFreeTimes) {
+      TreeMap<Integer, Integer> changes = new TreeMap<Integer, Integer>();
+      for (Map.Entry e : optionalAttendeesFreeTimes.entrySet()) {
+        for (TimeRange time : (ArrayList<TimeRange>) e.getValue()) {
+          changes.put(time.start(), changes.getOrDefault(time.start(), 0) + 1);
+          changes.put(time.end(), changes.getOrDefault(time.end(), 0) - 1);
         }
       }
-      if (isRelevant) {
-        relevantEvents.add(event);
+      return changes;
+    }
+
+    /**
+     * Updates optimalMeetingTimes based on current meeting time's attendance.
+     *
+     * @param time current meeting time range
+     */
+    private void updateOptimalTimes(TimeRange time) {
```

Comment:
What is this function supposed to be doing?

When I read it, I see that it can
(1) do nothing. The reason for doing nothing is opaque to me as a reader (`time.duration() < minTime` means almost nothing to me)

(2) inspect these random member variables and clear `optimalMeetingTimes`, which is also unclear to me

(3) inspect these values and add a member to optimalMeetingTimes.

This function should not be factored out. It makes no sense in isolation.

---

Separately, again, I wonder why this function returns nothing. Does the caller not care if optimalMeetingTimes is cleared out? This seems consequential.

Reply:
Got rid of it and moved it back into the for loop.